## Impressions and Reviews  
- preprocessing : sparse(nunique<2), duplicate한 feature drop
- lgbm, xgb, cat 을 blending (ensemble)



## Load Required Libraries

In [10]:
import numpy as np
import pandas as pd
import gc

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from sklearn import model_selection
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import RobustScaler

import lightgbm as lgb
import xgboost as xgb
from catboost import CatBoostRegressor

from IPython.display import display # Allows the use of display() for DataFrames

import warnings
warnings.filterwarnings('ignore')

## Load Train and Test Data

In [11]:
# Read train and test files
train_df = pd.read_csv('/content/train.csv')
test_df = pd.read_csv('/content/test.csv')

### Train Data

In [12]:
train_df.head()

ID      target  48df886f9  ...  fb36b89d9  7e293fbaf  9fc776466
0  000d6aaf2  38000000.0        0.0  ...          0          0          0
1  000fbd867    600000.0        0.0  ...          0          0          0
2  0027d6b71  10000000.0        0.0  ...          0          0          0
3  0028cbf45   2000000.0        0.0  ...          0          0          0
4  002a68644  14400000.0        0.0  ...          0          0          0

[5 rows x 4993 columns]

### Test Data

In [13]:
test_df.head()

ID  48df886f9  0deb4b6a8  ...  fb36b89d9  7e293fbaf  9fc776466
0  000137c73        0.0        0.0  ...        0.0        0.0        0.0
1  00021489f        0.0        0.0  ...        0.0        0.0        0.0
2  0004d7953        0.0        0.0  ...        0.0        0.0        0.0
3  00056a333        0.0        0.0  ...        0.0        0.0        0.0
4  00056d8eb        0.0        0.0  ...        0.0        0.0        0.0

[5 rows x 4992 columns]

## Train and Test Data Info

In [14]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4459 entries, 0 to 4458
Columns: 4993 entries, ID to 9fc776466
dtypes: float64(1845), int64(3147), object(1)
memory usage: 169.9+ MB


So there are a total of 4993 columns out of which 1845 are of type float64, 3147 are int64 and 1 is object (ID is the object column)

In [15]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49342 entries, 0 to 49341
Columns: 4992 entries, ID to 9fc776466
dtypes: float64(4991), object(1)
memory usage: 1.8+ GB


So there are a total of 4992 columns in the test set out of which 4991 are of type float64 and 1 is object (ID is the object column)

## Check for Missing Values

In [16]:
# train set 결측치 확인
print("Total Train Features with NaN Values = " + str(train_df.columns[train_df.isnull().sum() != 0].size))
if (train_df.columns[train_df.isnull().sum() != 0].size):
    print("Features with NaN => {}".format(list(train_df.columns[train_df.isnull().sum() != 0])))
    train_df[train_df.columns[train_df.isnull().sum() != 0]].isnull().sum().sort_values(ascending = False)

Total Train Features with NaN Values = 0


In [17]:
# test set 결측치 확인
print("Total Test Features with NaN Values = " + str(test_df.columns[test_df.isnull().sum() != 0].size))
if (test_df.columns[test_df.isnull().sum() != 0].size):
    print("Features with NaN => {}".format(list(test_df.columns[test_df.isnull().sum() != 0])))
    test_df[test_df.columns[test_df.isnull().sum() != 0]].isnull().sum().sort_values(ascending = False)

Total Test Features with NaN Values = 0


## Check and Remove Constant Features

In [18]:
# 값이 constant한 column 제거(std == 0) -> 256개 col 제거
colsToRemove = []
for col in train_df.columns:
    if col != 'ID' and col != 'target':
        if train_df[col].std() == 0: 
            colsToRemove.append(col)
        
# remove constant columns in the training set
train_df.drop(colsToRemove, axis=1, inplace=True)

# remove constant columns in the test set
test_df.drop(colsToRemove, axis=1, inplace=True) 

print("Removed `{}` Constant Columns\n".format(len(colsToRemove)))
print(colsToRemove)

Removed `256` Constant Columns

['d5308d8bc', 'c330f1a67', 'eeac16933', '7df8788e8', '5b91580ee', '6f29fbbc7', '46dafc868', 'ae41a98b6', 'f416800e9', '6d07828ca', '7ac332a1d', '70ee7950a', '833b35a7c', '2f9969eab', '8b1372217', '68322788b', '2288ac1a6', 'dc7f76962', '467044c26', '39ebfbfd9', '9a5ff8c23', 'f6fac27c8', '664e2800e', 'ae28689a2', 'd87dcac58', '4065efbb6', 'f944d9d43', 'c2c4491d5', 'a4346e2e2', '1af366d4f', 'cfff5b7c8', 'da215e99e', '5acd26139', '9be9c6cef', '1210d0271', '21b0a54cb', 'da35e792b', '754c502dd', '0b346adbd', '0f196b049', 'b603ed95d', '2a50e001c', '1e81432e7', '10350ea43', '3c7c7e24c', '7585fce2a', '64d036163', 'f25d9935c', 'd98484125', '95c85e227', '9a5273600', '746cdb817', '6377a6293', '7d944fb0c', '87eb21c50', '5ea313a8c', '0987a65a1', '2fb7c2443', 'f5dde409b', '1ae50d4c3', '2b21cd7d8', '0db8a9272', '804d8b55b', '76f135fa6', '7d7182143', 'f88e61ae6', '378ed28e0', 'ca4ba131e', '1352ddae5', '2b601ad67', '6e42ff7c7', '22196a84c', '0e410eb3d', '992e6d1d3', '90a7

## Remove Duplicate Columns

In [36]:
groups = train_df.columns.to_series().groupby(train_df.dtypes).groups
groups

{int64: ['0deb4b6a8', 'a8cb14b00', '2f0771a37', '30347e683', 'd08d1fbe3', '6ee66e115', '77c9823f2', '8d6c2a0b2', '4681de4fd', 'adf119b9a', '96f83a237', '6c7a4567c', '4fcfd2b4d', '71cebf11c', 'd966ac62c', 'c88d108c9', 'ff7b471cd', 'bc3f77679', '0eff5bf95', '2cb4d123e', '316b978cd', '5d5c5ce6d', 'ec863cb52', '69d8b4020', 'b0868a049', 'c36bcacc5', '151d318cd', 'ab0cddb90', 'c87722678', 'a8c320153', 'a04684f1f', '645b47cde', '7298ca1ef', 'd80abf8bc', '654dd8a3b', '7fc39583c', 'e421c414e', '964335fdf', '5a86cabd0', '6cb207ac9', 'c1bc828da', 'e90ed19da', 'c4d38135f', '968b01ca3', '26b423c42', '0b8e10df6', '603e85eb7', '605728195', 'ee7e4581d', 'a939664aa', 'a72e0bf30', 'e662f1672', '78a879b5d', 'bd550871c', '3268914c7', 'ec73030f6', 'c5b30277a', 'c16a7d537', 'bd6671d31', 'aab0aeb4b', '9a9b4a012', 'c377b9acf', 'f3e544d66', '80c0c9910', '9f8df01bd', '563d2f09a', '7ee833549', '7b58c8e37', 'cde9c35e8', '1e8801477', '3770cb9fa', '133714358', '15dba7177', '604ac0633', 'cbf34125e', '4563c7ee5', 'c6

In [20]:
# remove duplicate columns in the training set
train_df.drop(colsToRemove, axis=1, inplace=True) 

# remove duplicate columns in the testing set
test_df.drop(colsToRemove, axis=1, inplace=True)

print("Removed `{}` Duplicate Columns\n".format(len(colsToRemove)))
print(colsToRemove)

Removed `5` Duplicate Columns

['34ceb0081', '8d57e2749', '168b3e5bc', 'a765da8bc', 'acc5b709d']


In [19]:
# 중복 column 제거
%%time
def duplicate_columns(frame):

    # dtype 별 col 을 key, value로 하는 dict 생성
    groups = frame.columns.to_series().groupby(frame.dtypes).groups
    dups = []

    for t, v in groups.items():

        cs = frame[v].columns
        vs = frame[v]
        lcs = len(cs)

        for i in range(lcs):
            ia = vs.iloc[:,i].values
            for j in range(i+1, lcs):
                ja = vs.iloc[:,j].values
                if np.array_equal(ia, ja): # dtype 같은 두개 col이 실제로 값들이 같은지 확인
                    dups.append(cs[i])
                    break

    return dups

colsToRemove = duplicate_columns(train_df)
print(colsToRemove)

['34ceb0081', '8d57e2749', '168b3e5bc', 'a765da8bc', 'acc5b709d']
CPU times: user 5min 11s, sys: 98.2 ms, total: 5min 11s
Wall time: 5min 12s


## Drop Sparse Data

In [21]:
# sparse한 feature drop (nunique==1)
def drop_sparse(train, test):
    flist = [x for x in train.columns if not x in ['ID','target']]
    for f in flist:
        if len(np.unique(train[f]))<2:
            train.drop(f, axis=1, inplace=True)
            test.drop(f, axis=1, inplace=True)
    return train, test

In [22]:
%%time
train_df, test_df = drop_sparse(train_df, test_df)

CPU times: user 552 ms, sys: 33 µs, total: 552 ms
Wall time: 553 ms


In [23]:
gc.collect()
print("Train set size: {}".format(train_df.shape))
print("Test set size: {}".format(test_df.shape))

Train set size: (4459, 4732)
Test set size: (49342, 4731)


## Build Train and Test Data for Modeling

In [24]:
# X는 ID 제외, Y는 log1p 변환
X_train = train_df.drop(["ID", "target"], axis=1)
y_train = np.log1p(train_df["target"].values)

X_test = test_df.drop(["ID"], axis=1)

In [25]:
# 8:2 로 split
dev_X, val_X, dev_y, val_y = train_test_split(X_train, y_train, test_size = 0.2, random_state = 42)

## LightGBM

In [26]:
def run_lgb(train_X, train_y, val_X, val_y, test_X):
    params = {
        "objective" : "regression",
        "metric" : "rmse",
        "num_leaves" : 40,
        "learning_rate" : 0.004,
        "bagging_fraction" : 0.6,
        "feature_fraction" : 0.6,
        "bagging_frequency" : 6,
        "bagging_seed" : 42,
        "verbosity" : -1,
        "seed": 42
    }
    
    lgtrain = lgb.Dataset(train_X, label=train_y)
    lgval = lgb.Dataset(val_X, label=val_y)
    evals_result = {}
    model = lgb.train(params, lgtrain, 5000, 
                      valid_sets=[lgtrain, lgval], 
                      early_stopping_rounds=100, 
                      verbose_eval=150, 
                      evals_result=evals_result)
    
    pred_test_y = np.expm1(model.predict(test_X, num_iteration=model.best_iteration))
    return pred_test_y, model, evals_result

In [27]:
# Training LGB
pred_test, model, evals_result = run_lgb(dev_X, dev_y, val_X, val_y, X_test)
print("LightGBM Training Completed...")

Training until validation scores don't improve for 100 rounds.
[150]	training's rmse: 1.50793	valid_1's rmse: 1.53931
[300]	training's rmse: 1.34402	valid_1's rmse: 1.46602
[450]	training's rmse: 1.23291	valid_1's rmse: 1.43402
[600]	training's rmse: 1.14927	valid_1's rmse: 1.41918
[750]	training's rmse: 1.08343	valid_1's rmse: 1.41347
[900]	training's rmse: 1.03011	valid_1's rmse: 1.41164
[1050]	training's rmse: 0.985188	valid_1's rmse: 1.4116
Early stopping, best iteration is:
[976]	training's rmse: 1.00631	valid_1's rmse: 1.41125
LightGBM Training Completed...


In [28]:
# feature importance (gain은 %)
print("Features Importance...")
gain = model.feature_importance('gain')
featureimp = pd.DataFrame({'feature':model.feature_name(), 
                   'split':model.feature_importance('split'), 
                   'gain':100 * gain / gain.sum()}).sort_values('gain', ascending=False)
print(featureimp[:50])

Features Importance...
        feature  split      gain
4130  f190486d6    803  9.017404
2375  58e2e02e6    715  5.294145
3465  eeb9cd3aa    675  4.294044
4020  15ace8c9f    541  3.095043
2614  9fd594eec    379  2.850291
8     20aa07010    409  2.112768
834   6eef030c1    342  1.400379
3571  58232a6fb    380  1.400253
1457  b43a7cfd5    399  1.187425
3661  491b9ee45    297  1.048269
2687  fb0f5dbfe    410  0.997311
1482  024c577b9    258  0.931900
3867  2288333b4    195  0.860454
2079  58e056e12    347  0.857672
4185  f74e8f13d    352  0.849962
4343  1702b5bf0    291  0.836605
4508  c47340d97    313  0.815153
828   6786ea46d    184  0.798005
566   66ace2992    276  0.778013
3722  d6bb78916    313  0.770098
3791  ed8ff54b5    169  0.764647
3220  ced6a7e91    240  0.720281
4028  5c6487af1    180  0.712760
3886  50e4f96cf    152  0.686449
863   fc99f9426    238  0.667129
1378  6cf7866c1    160  0.621800
34    87ffda550    162  0.604544
853   bc70cbc26    139  0.599770
3811  adb64ff71    2

## XGB Modeling

In [29]:
def run_xgb(train_X, train_y, val_X, val_y, test_X):
    params = {'objective': 'reg:linear', 
          'eval_metric': 'rmse',
          'eta': 0.001,
          'max_depth': 10, 
          'subsample': 0.6, 
          'colsample_bytree': 0.6,
          'alpha':0.001,
          'random_state': 42, 
          'silent': True}
    
    tr_data = xgb.DMatrix(train_X, train_y)
    va_data = xgb.DMatrix(val_X, val_y)
    
    watchlist = [(tr_data, 'train'), (va_data, 'valid')]
    
    model_xgb = xgb.train(params, tr_data, 2000, watchlist, maximize=False, early_stopping_rounds = 100, verbose_eval=100)
    
    dtest = xgb.DMatrix(test_X)
    xgb_pred_y = np.expm1(model_xgb.predict(dtest, ntree_limit=model_xgb.best_ntree_limit))
    
    return xgb_pred_y, model_xgb

In [30]:
# Training XGB
pred_test_xgb, model_xgb = run_xgb(dev_X, dev_y, val_X, val_y, X_test)
print("XGB Training Completed...")

[0]	train-rmse:14.0877	valid-rmse:14.0768
Multiple eval metrics have been passed: 'valid-rmse' will be used for early stopping.

Will train until valid-rmse hasn't improved in 100 rounds.
[100]	train-rmse:12.7683	valid-rmse:12.756
[200]	train-rmse:11.5765	valid-rmse:11.5629
[300]	train-rmse:10.5	valid-rmse:10.4854
[400]	train-rmse:9.52802	valid-rmse:9.51304
[500]	train-rmse:8.65065	valid-rmse:8.6361
[600]	train-rmse:7.85865	valid-rmse:7.84473
[700]	train-rmse:7.14388	valid-rmse:7.13201
[800]	train-rmse:6.49911	valid-rmse:6.48993
[900]	train-rmse:5.91691	valid-rmse:5.91106
[1000]	train-rmse:5.3925	valid-rmse:5.39137
[1100]	train-rmse:4.9196	valid-rmse:4.92423
[1200]	train-rmse:4.49343	valid-rmse:4.50457
[1300]	train-rmse:4.10977	valid-rmse:4.12874
[1400]	train-rmse:3.76474	valid-rmse:3.79266
[1500]	train-rmse:3.45493	valid-rmse:3.49292
[1600]	train-rmse:3.17654	valid-rmse:3.22512
[1700]	train-rmse:2.92698	valid-rmse:2.98712
[1800]	train-rmse:2.70352	valid-rmse:2.77675
[1900]	train-rmse:

## Catboost

In [31]:
cb_model = CatBoostRegressor(iterations=500,
                             learning_rate=0.05,
                             depth=10,
                             eval_metric='RMSE',
                             random_seed = 42,
                             bagging_temperature = 0.2,
                             od_type='Iter',
                             metric_period = 50,
                             od_wait=20)

In [32]:
cb_model.fit(dev_X, dev_y,
             eval_set=(val_X, val_y),
             use_best_model=True,
             verbose=50)

0:	learn: 1.7518683	test: 1.6878429	best: 1.6878429 (0)	total: 3.37s	remaining: 28m
50:	learn: 1.4789181	test: 1.5197196	best: 1.5197196 (50)	total: 2m 33s	remaining: 22m 34s
100:	learn: 1.3788775	test: 1.4780503	best: 1.4780503 (100)	total: 5m 4s	remaining: 20m 1s
150:	learn: 1.3203081	test: 1.4647232	best: 1.4647232 (150)	total: 7m 34s	remaining: 17m 29s
200:	learn: 1.2546056	test: 1.4505032	best: 1.4504536 (198)	total: 10m 4s	remaining: 14m 58s
250:	learn: 1.1817435	test: 1.4381513	best: 1.4381513 (250)	total: 12m 34s	remaining: 12m 28s
300:	learn: 1.1208241	test: 1.4320748	best: 1.4314980 (297)	total: 15m 5s	remaining: 9m 58s
350:	learn: 1.0770301	test: 1.4298988	best: 1.4297384 (347)	total: 17m 38s	remaining: 7m 29s
Stopped by overfitting detector  (20 iterations wait)

bestTest = 1.428228031
bestIteration = 376

Shrink model to first 377 iterations.


In [33]:
pred_test_cat = np.expm1(cb_model.predict(X_test))

## Combine Predictions

In [ ]:
# lgb:xgb:cat = 5:3:2 로 blending해서 pred값 결정
sub = pd.read_csv('../input/sample_submission.csv')

sub_lgb = pd.DataFrame()
sub_lgb["target"] = pred_test

sub_xgb = pd.DataFrame()
sub_xgb["target"] = pred_test_xgb

sub_cat = pd.DataFrame()
sub_cat["target"] = pred_test_cat

sub["target"] = (sub_lgb["target"] * 0.5 + sub_xgb["target"] * 0.3 + sub_cat["target"] * 0.2)

In [ ]:
print(sub.head())
sub.to_csv('sub_lgb_xgb_cat.csv', index=False)